Implemented logging to Graylog server

Faldon opened the pull request:
# For an old feature request, I implemented logging as GELF to a Graylog server.
## Features
+ Configurable to use tcp or udp in system config
+ Auto-chunking udp messages larger than 1024 bytes
+ Log server and port configurable in host:port notation
## Implementation
Added a new class handling writing the logs to the configured server through a socket connection. If the connection fails, it returns silently.
GELF supports chunking of UDP messages up to 128 chunks. So a log message can be approx. 126kB in size or it is silently discarded.
I also added the ability to configure the logging method through the occ log:manage command scope.

Closes #95

Faldon committed:
Implemented logging to graylog server.
Added the appropriate logging class.
Provided the necessary system config examples.
Added loading the log class from the LogFactory.
Extended the LogFactory unit tests to include the graylog class.
Added Graylog test class skeleton.

Signed-off-by: Thomas Pulzer <t.pulzer@thesecretgamer.de>

Faldon committed:
Reformatted and rearranged some code parts.
Changed method of how to connect to remote server for writing logs.
Wrote unit tests for chunked udp, unchunked udp and tcp connections.

Signed-off-by: Thomas Pulzer <t.pulzer@thesecretgamer.de>

Faldon committed:
Renamed config key for Graylog protocol to use.
Fixed wrong byte packing when sending chunked UDP packets.
Improved test code and comments.

Signed-off-by: Thomas Pulzer <t.pulzer@thesecretgamer.de>

Faldon committed:
Added log backend graylog to occ log:manage command.

Signed-off-by: Thomas Pulzer <t.pulzer@thesecretgamer.de>

Faldon committed:
Added ability to configure graylog remote host and connection protocol with occ

Signed-off-by: Thomas Pulzer <t.pulzer@thesecretgamer.de>

Faldon committed:
Added test of the log:manage method if graylog backend is set.

Signed-off-by: Thomas Pulzer <t.pulzer@thesecretgamer.de>

kesselb commented on lib/private/Log/Graylog.php:
```diff
@@ -0,0 +1,107 @@
+	 */
+	public function write(string $app, $message, int $level) {
+		$chunks = [];
+		$msg = '{"version":"' . self::$VERSION . '","host":"' .
```
I think you could use a array like `json_encode(['version' => self::$VERSION])` here

Faldon replied:
Oh dear. :man_facepalming: Sometimes I don't see the wood for the trees. 
That also eliminates the manual replacement of the line break character (and probably more vulnerabilities with regards to handling message input).
That's an excellent change I will implement.

Faldon committed:
Dropped constructing the GELF message manually in favour for json_encode
This revealed former incorrect encoding of field 'level' (is nuber, but
was encoded as string. Therefore, the test method had to be adjusted as
well.

Signed-off-by: Thomas Pulzer <t.pulzer@thesecretgamer.de>

Faldon commented:
Considering the CI check... Shall / must I rebase my branch onto master?

kesselb commented:
> Considering the CI check... Shall / must I rebase my branch onto master?

The autoloaders are not up to date
Please run: bash build/autoloaderchecker.sh
And commit the result

Faldon committed:
Updated autoloaders.

Signed-off-by: Thomas Pulzer <t.pulzer@thesecretgamer.de>

Faldon commented:
Rebased onto master to have all the checks passed.

kesselb commented on core/Command/Log/Manage.php:
```diff
@@ -126,6 +160,17 @@ protected function validateBackend($backend) {
+	 * @throws \InvalidArgumentException
+	 */
+	protected function isBackendGraylogSet($new=null) {
+		$old = $this->config->getSystemValue('log_type', self::DEFAULT_BACKEND);
```
I'm not sure about this :thinking: It's not possible to set `greylog_host` or `greylog_proto` if the current `log_type` or the new `log_type` is not greylog. That makes sense but is very strict. There is no harm if `greylog_host` or `greylog_proto` are set with another `log_type`. I think we need some input from another reviewer.

kesselb commented on core/Command/Log/Manage.php:
```diff
@@ -99,6 +112,20 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 		}
 
+		if ($host = $input->getOption('host')) {
+			array_key_exists('log_type', $toBeSet) ?
```
I would like to suggest to change this a bit. Its quite hard to read and we have the same logic two times. The `isBackendGraylogSet` method should accept an array $toBeSet and do the array_key_exists check.

kesselb commented on lib/private/Log/Graylog.php:
```diff
@@ -0,0 +1,109 @@
+	public function __construct(IConfig $config) {
+		$this->host = gethostname();
+		$this->protocol = $config->getSystemValue('graylog_proto', 'udp');
+		$address = $config->getSystemValue('graylog_host', '');
```
Code is fine. I would pick a slightly different approach but everyone has a different code style so this is up to you.

```
		$address = explode(':', $config->getSystemValue('graylog_host'));
		if (count($address) === 2) {
			$this->target = $address[0];
			$this->port = (int)$address[1];
		} else {
			$this->target = $address[0];
			$this->port = 514;
		}
```

kesselb commented on core/Command/Log/Manage.php:
```diff
@@ -126,6 +160,17 @@ protected function validateBackend($backend) {
+	 * @param string $new
+	 * @throws \InvalidArgumentException
+	 */
+	protected function isBackendGraylogSet($new=null) {
```
We should find another name for it.

kesselb reviewed (commented):
Thanks for the great contribution :+1:  Please excuse the late code review :disappointed:

MorrisJobke commented on tests/lib/Log/GraylogTest.php:
```diff
@@ -0,0 +1,147 @@
+	}
+
+	protected function tearDown() {
+		if (isset($this->protocol_restore)) {
```
I would keep it, because then the proper "is everything set as wanted and also set together via config" covered.

kesselb replied:
> is everything set as wanted and also set together via config

=> Integration test. I see your point but there is a good chance that in the future this unit test on drugs will fail :sunglasses:

MorrisJobke commented on lib/private/Log/Graylog.php:
```diff
@@ -0,0 +1,109 @@
+		$msg = json_encode([
+			'version' => self::$VERSION,
+			'host' => $this->host,
+			'short_message' => '{'.$app.'} '.$message,
```
You might want to use the new way here as well:

```suggestion
			'short_message' => $this->logDetailsAsJSON($app, $message, $level),
```

Was added just now via #16416

MorrisJobke replied:
maybe the non-encoded one is better here: `logDetails()`, because it does not return a string with JSON, but an object with keys.

MorrisJobke replied:
cc @juliushaertl

juliushaertl replied:
Yes, if Graylog can handle nested data in the short_message key that should be the way to go.

rullzer commented:
Lets merge this early in 18.
@juliushaertl could you help out since you touched the logging code recently

ChristophWurst commented:
Lot's of unaddressed review comments. Some unanswered questions. This needs triage. Not assigning to 19 because pushing this from release to release has no point.

ChristophWurst commented:
What might make more sense IMO is to use some existing packages like https://packagist.org/packages/giacomofurlan/php-graylog-gelf in combination with #18200. And preferably make this an app.

Like
* Use PSR for the logger interface
* Make the logger implementation pluggable
* Create an implementation for Graylog as an app

ChristophWurst commented:
Let's close this due to lack of activity. If anyone would like to bring this feature to life, please do so as an app as described in my previous comment.
